File: backend/mr-db-api/src/main/java/io/featurehub/db/api/EnvironmentApi.java
package io.featurehub.db.api;

import io.featurehub.mr.model.Application;
import io.featurehub.mr.model.ApplicationRoleType;
import io.featurehub.mr.model.Environment;
import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.Portfolio;
import io.featurehub.mr.model.RoleType;
import io.featurehub.mr.model.SortOrder;

import java.util.List;
import java.util.Set;
import java.util.UUID;

public interface EnvironmentApi {
  EnvironmentRoles personRoles(Person current, UUID eid);

  List<Environment> setOrdering(Application app, List<Environment> environments);

  class DuplicateEnvironmentException extends Exception {}
  class InvalidEnvironmentChangeException extends Exception {}
  boolean delete(UUID id);
  Environment get(UUID id, Opts opts, Person current);
  Environment update(UUID envId, Environment env, Opts opts) throws OptimisticLockingException, DuplicateEnvironmentException, InvalidEnvironmentChangeException;
  Environment create(Environment env, Application app, Person whoCreated) throws DuplicateEnvironmentException, InvalidEnvironmentChangeException;
  List<Environment> search(UUID appId, String filter, SortOrder order, Opts opts, Person current);
  Portfolio findPortfolio(UUID envId);
}

File: backend/mr-db-api/src/main/kotlin/io/featurehub/db/api/EnvironmentApi.kt
package io.featurehub.db.api

import io.featurehub.mr.model.*
import java.util.*

interface EnvironmentApi {
  fun portfolioEnvironmentBelongsTo(eId: UUID): UUID?
  fun personRoles(current: Person, eid: UUID): EnvironmentRoles?
  fun setOrdering(app: Application, environments: List<Environment>): List<Environment>?
  class DuplicateEnvironmentException : Exception()
  class InvalidEnvironmentChangeException : Exception()

  fun delete(id: UUID?): Boolean
  operator fun get(id: UUID, opts: Opts?, current: Person?): Environment?

  @Throws(
    OptimisticLockingException::class,
    DuplicateEnvironmentException::class,
    InvalidEnvironmentChangeException::class
  )
  fun update(envId: UUID, env: Environment, opts: Opts): Environment?
  fun update(application: UUID, env: UpdateEnvironmentV2, opts: Opts): Environment?

  @Throws(DuplicateEnvironmentException::class, InvalidEnvironmentChangeException::class)
  fun create(env: CreateEnvironment, appId: UUID, whoCreated: Person): Environment?
  fun search(
    appId: UUID?,
    filter: String?,
    order: EnvironmentSortOrder?,
    opts: Opts?,
    current: Person?
  ): List<Environment>

  fun findPortfolio(envId: UUID?): Portfolio?
  fun unpublishEnvironments(appId: UUID, environments: List<UUID>?): Int
  fun updateEnvironment(eid: UUID, env: UpdateEnvironment, opts: Opts): Environment?

  /**
   * If null, then no app exists or there are no environments
   * If empty, all environments, otherwise environment list
   */
  fun getEnvironmentsUserCanAccess(appId: UUID, person: UUID): List<UUID>?
}
